plugins/plugin-docker/che-plugin-docker-machine/src/main/java/org/eclipse/che/plugin/docker/machine/cleaner/DockerContainerCleaner.java

```diff
@@ -0,0 +1,123 @@
+import org.eclipse.che.plugin.docker.client.json.ContainerListEntry;
+import org.eclipse.che.plugin.docker.machine.DockerContainerNameGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import static org.eclipse.che.plugin.docker.machine.DockerContainerNameGenerator.ContainerNameInfo;
+
+/**
+ * Job for periodically clean up inactive docker containers
+ *
+ * @author Alexander Andrienko
+ */
+@Singleton
+public class DockerContainerCleaner implements Runnable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DockerContainerCleaner.class);
+
+    private final MachineManager               machineManager;
+    private final DockerConnector              dockerConnector;
+    private final DockerContainerNameGenerator nameGenerator;
+
+    @Inject
+    public DockerContainerCleaner(MachineManager machineManager,
+                                  DockerConnector dockerConnector,
+                                  DockerContainerNameGenerator nameGenerator) {
+        this.machineManager = machineManager;
+        this.dockerConnector = dockerConnector;
+        this.nameGenerator = nameGenerator;
+    }
+
+    @ScheduleRate(periodParameterName = "machine.docker.unused_containers_cleanup_period_min",
+                  initialDelayParameterName = "machine.docker.unused_containers_cleanup_period_min",
+                  unit = TimeUnit.MINUTES)
+    @Override
+    public void run() {
+        try {
+            List<ContainerListEntry> allDockerContainers = dockerConnector.listContainers();
+            List<MachineImpl> machines = machineManager.getMachines();
+
+            List<ContainerListEntry> unUsedContainers = findUnUsedContainers(allDockerContainers, machines);
+
+            for (ContainerListEntry container : unUsedContainers) {
+                killContainer(container);
+                removeContainer(container);
+            }
+        } catch (IOException e) {
+            LOG.error("Failed to get list docker containers", e);
+        } catch (MachineException e) {
+            LOG.error("Failed to get list machines to clean up unused containers", e);
+        } catch (Exception e) {
+            LOG.error("Failed to clean up inactive containers", e);
+        }
+    }
+
+    private List<ContainerListEntry> findUnUsedContainers(List<ContainerListEntry> containers, List<MachineImpl> machines) {
+        List<ContainerListEntry> unUsedContainers = new ArrayList<>();
+        for (ContainerListEntry container : containers) {
+            Optional<ContainerNameInfo> optional = nameGenerator.parse(container.getNames()[0]);
+            if (!optional.isPresent()) {
+                continue;
+            }
+            final ContainerNameInfo containerNameInfo = optional.get();
+            boolean containerIsUsed = machines.stream()
```

Comment:
For example you have 2k machine and 2k + 1 containers. You try to find machine for container which is not used by machines anymore. You have to perform ~ 2k * 2k = 4kk opertaions.